### PR TITLE
add `zlib-rs` support via the `libz-rs-sys` C api for `zlib-rs`

### PR DESCRIPTION
**Pull request by folkertdev**
This PR adds (experimental) support for using flate2 with the zlib-rs backend. The `zlib-rs` crate is a pure-rust implementation of the zlib api that makes use of SIMD to achieve performance close to zlib-ng (currently 5% to 10% slower).

- [ISRG project page](https://www.memorysafety.org/initiative/zlib/)
- [`libz-rs-sys` docs](https://docs.rs/libz-rs-sys/0.1.0/libz_rs_sys/index.html)

For the moment, the libz-rs-sys C api is used for the integration. At some point in the future, zlib-rs will get a more convenient rust api that will require less (hopefully no) unsafe code.

cc @joshtriplett

### CODE REVIEW

File: src/ffi/c.rs
Comment (folkertdev) [inline src/ffi/c.rs:74]: the trick is that because zlib-rs can do equality comparisons on the function pointer, it can know when its built-in allocator is used, and this allocator guarantees 64-byte alignment. The allocation of `zfree` and `zalloc` below is not needed (and that implementation is less efficient to deal with allignment).

File: src/ffi/c.rs
Comment (folkertdev) [inline src/ffi/c.rs:90]: in the case of zlib-rs, the allocator logic is unused. That is an error for this crate, so it's only conditionally compiled now.

File: src/ffi/c.rs
Comment (folkertdev) [inline src/ffi/c.rs:106]: The only two occurences of this `AllocSize` type are in the signature of `zalloc`, which is a private function. Can this type alias be inlined into the signature?

File: src/ffi/c.rs
Comment (folkertdev) [inline src/ffi/c.rs:449]: zlib-rs has a `const fn zlibVersion() -> *const c_char`. So maybe these constants could all have type `*const c_char`? the only downside I see is that that is harder to print for debugging, but because these are internal constants that seems fine.

File: .github/workflows/main.yml
Comment (folkertdev) [inline .github/workflows/main.yml:47]: we should eventually be able to compile on mingw, but the current release doesn't, and we don't test for it, so disabling for now.

File: src/ffi/c.rs
Comment (Byron) [inline src/ffi/c.rs:104]: What's the motivation for the change from `AllocSize` to `uInt`?

I see this was previously set to `pub type AllocSize = libz::uInt;`, and the `*` imports don't make it easier to see where the type is actually coming from.

File: src/ffi/c.rs
Comment (Byron) [reply src/ffi/c.rs:447]: It would certainly be nice to automatically obtain the correct value for this, as it would be my guess that these constants quickly go stale otherwise. I also don't know how important these values are, assuming they might be printed under certain circumstances.

Until that's clear, it's probably best to integrate with what's there as is done here.

**Review by Byron — COMMENTED** [review]
Thanks a lot for contributing what could one day be the new default backend, and replace `minizoxide` in this space.

As CI is failing on Windows, I think that might be a good reason to run tests on Windows (and maybe more) platforms as well.

Further, when looking at the dependencies of `libz-rs-sys`, I was surprised that `zlib-rs` [has what looks like test-dependencies](https://github.com/memorysafety/zlib-rs/blob/main/zlib-rs/Cargo.toml#L21) as part of its normal dependencies. It's clear that these are optional and behind features, but I would have thought that `dev-dependencies` work fine for unit-tests as well. It somehow worried me that this (what seems like a) workaround was necessary, and I hoped there is another way. Maybe APIs could be made public behind a hidden feature toggle, and a separate test-only crate could carry all the dependencies? Then the separation would be clear, and maybe some test-specific code could be extracted into a separate space. This comment is probably not relevant to this PR though, merely a side-note.

I would definitely wait for CI to be fixed, and a second reviewer.

File: src/ffi/c.rs
Comment (folkertdev) [reply src/ffi/c.rs:447]: it's important that the major version is the same, (that is what is checked), so the value won't go stale that quickly. Also tests will fail hard when that happens, so there is no real risk of the value going state on the main branch here.

File: src/ffi/c.rs
Comment (folkertdev) [reply src/ffi/c.rs:104]: yeah, `AllocSize` was only used in this signature (turns out) and I didn't think that the alias really had any value at that point.

**Comment by folkertdev** [comment]
thanks for the review

- we're working on windows support (already merged into our `main` branch but needs a little cleanup still). Should be ready soon.
- the dependency situation is kind of nasty. From what I observe, it seems that if crate A has a dev-dependency on crate B, then crate B is not compiled with `cfg(test)` to be true. Our crate B needs to expose/derive things for testing in crate A. So, we haven't yet found a better solution than the feature flags, but it is far from ideal, I'd really like to not be doing that.
- we've been in regular contact with @joshtriplett so he'd be a natural second reviewer here.

**Comment by Byron** [comment]
My apologies, that callout what unnecessary, and lacked the appreciation that is due for it as the backend which *just works*. I got carried away by the pure-Rust-powered zlib-ng like performance as default here, it's something that `gitoxide` benefits from a lot.

But maybe this could be a starting point for eventually bringing both projects together (*without having any idea if this is technically or even legally possible*).

**Comment by oyvindln** [comment]
Ah sorry, maybe I came across as overly grumpy, wasn't meant for you or anyone in particular, it's just happened a few times now

**Comment by Byron** [comment]
Right, thanks for clarifying this! If miniz-oxide is 100% safe Rust, there is value in that alone, independently of performance. Then, even if `zlib-rs` is equivalent functionally and regarding portability, figuring out what the default backend should be wouldn't be straightforward anymore and probably stay as is :).

**Comment by folkertdev** [comment]
We certainly have no ill will versus miniz-oxide: it's simple, small, consistent between platforms, no unsafe besides some C api stuff, and performs slightly better than the original zlib, which is perfectly adequate for many use cases.

`zlib-rs` is more ambitious though: it aims to match zlib-ng performance and properly provide the full zlib api, meaning that we can create a drop-in replacement for the zlib dynamic library.

I'm not sure a generic, shared deflate implementation works, but we're happy to share code/knowlegde. It is our impression however that the majority of performance improvements stem from the explicit use of SIMD instructions, which miniz-oxide does not seem to want to use?

---

We believe to have fixed the windows issues. I'll try that here on CI now by setting the dependency to a git branch again. This should not be merged in that state. I'll report back after we've pushed another release to crates.io and think this is ready.

**Comment by oyvindln** [comment]
>It is our impression however that the majority of performance improvements stem from the explicit use of SIMD instructions, which miniz-oxide does not seem to want to use?

We don't want to use any unsafe code in miniz_oxide (external C api requires it of course but that's a separate thing), and as far as I know at least there isn't any way to use explicit SIMD without unsafe on stable rust (the [portable simd](https://doc.rust-lang.org/std/simd/index.html) API could help solve that but it seems to have been stalled as of now). There may be some crates out there that implement something similar though those again will be using unsafe internally.

To what extent explicit SIMD is needed to reach performance parity and to what extent it can be achieved with autovectorization with the right code layout I don't know.

In any case zlib-rs will now provide a rust alternative to using the C implementation of zlib-ng that's close to the same performance for users that are okay with the safety tradeoffs (and that hopefully won't be patched out by debian maintainers like [zlib-ng is](https://salsa.debian.org/rust-team/debcargo-conf/-/blob/master/src/flate2/debian/patches/disable-zlib-ng.patch))

**Comment by folkertdev** [comment]
The explicit SIMD is essential, e.g. https://github.com/memorysafety/zlib-rs/blob/main/zlib-rs/src/deflate/compare256.rs#L170, LLVM just can't see that idea with the wide comparison and bit counting.

There are a couple of tricks around breaking dependency chains in the bit reader that you may benefit from. E.g. https://github.com/Frommi/miniz_oxide/blob/master/miniz_oxide/src/inflate/core.rs#L423 the `if` here is very prone to misprediction. The optimized implementations just always add more bits to the buffer, e.g. we https://github.com/memorysafety/zlib-rs/blob/main/zlib-rs/src/inflate/bitreader.rs#L103 have no branches in that function at all. A bounds check would be acceptable because it is easy to predict.

What's the story behind removing `zlib-ng` from the `flate2` debian package?

**Comment by folkertdev** [comment]
We have released version 0.1.1 that fixes various issues on windows and macos (x86 and aarch64). These targets are now tested in our CI so there should be no regressions. 

From my side this PR is now done and ready to be merged

**Review by Byron — APPROVED** [review]
Thanks a lot for all your hard work and for seeing it through!

The only concern I have is that this CI is testing more platforms than the `zlib-rs` CI (even though Windows checks were added now), and I hope that this isn't prone to unexpected breakage here.

Then, I will leave it for @joshtriplett to finalize the review and possibly merge as well.

**Comment by folkertdev** [comment]
What platforms do we miss? mingw is deliberately omitted for now. I guess we could add a build with the beta target?

https://github.com/memorysafety/zlib-rs/blob/main/.github/workflows/checks.yaml#L25

**Comment by Byron** [comment]
Actually, I confused this repo with the one from `libz` which builds on many, many different platforms. Here it's indeed only `mingw` that was omitted. I think my argument *could* still apply, but much more weakly.

I don't think that nightly should play a role as nightly shouldn't ever break anything (and if it does, it's a bug that will be fixed before it hits stable).

**Review by joshtriplett — APPROVED** [review]
Looks good to me! Wrote a comment about the feature usage, but I don't think that's a blocker to merging.

File: Cargo.toml
Comment (folkertdev) [reply Cargo.toml:22]: yes exactly. it also clarifies that we're explicitly using the rust allocator; that will always be the default when the crate is used from rust, but not when we build a C dynamic library.

**Comment by folkertdev** [comment]
Thanks! would it be possible to make a release with these changes?

**Comment by Byron** [comment]
Yes, cold you create a PR that bumps the version? I can't self-approve my own PRs hence the need for help :).

**Comment by Byron** [comment]
A new release is now available: https://github.com/rust-lang/flate2-rs/releases/tag/1.0.29

**Comment by torokati44** [comment]
Does the README not mention this because it's still experimental (and therefore shouldn't be used "in production")?

**Comment by Byron** [comment]
That's a good point - it might have been an oversight, and it should probably be mentioned along with a note of being not necessarily production-proven yet.

**Pull request merged**
